Support bootstrap/byron witnesses in `evaluateMinimumFee`

Anviking opened the pull request:
- [x] Support bootstrap witnesses in `Cardano.Wallet.Write.Tx.evaluateMinimumFee`
- [x] Support bootstrap witnesses in the temporarily remaining `Cardano.Wallet.Shelley.Transactions.evaluateMinimumFee`
- [x] Re-use the exact same size estimation as the underlying coin-selection
- [x] Test that this size estimation doesn't underestimate w.r.t. the actual serialised size of the boot wits

### Comments

```
    bootstrap witnesses
      coin-selection's size estimation == balanceTx's size estimation [✔] (1697ms)
        +++ OK, passed 100 tests.
      balanceTx's size estimation >= measured serialized size [✔] (1824ms)
        +++ OK, passed 100 tests.

        overestimation/wit (100 in total):
        48% 9
        30% 2
         7% 4
         3% 0 (but with no wits)
         3% 5
         3% 6
         2% 11
         2% 8
         1% 10
         1% 7
```

### Issue Number

[ADP-2613](https://input-output.atlassian.net/browse/ADP-2613)

<!-- Reference the Jira/GitHub issue that this PR relates to, and which requirements it tackles.
  Note: Jira issues of the form ADP- will be auto-linked. -->

Anviking committed:
Support bootstrap witnesses in evaluateMinimumFee

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -2174,117 +2178,134 @@ estimateTxSize era skeleton =
+
+-- A Blake2b-224 hash, resulting in a 28-byte digest wrapped in CBOR, so
+-- with 2 bytes overhead (length <255, but length > 23)
+sizeOf_Hash28 :: Integer
```
All these are the exactly the same, just now usable in the entire module

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,80 @@ prop_balanceTransactionExistingReturnCollateral
             e -> counterexample (show e) False
 
+prop_bootstrapWitnesses
+    :: (Word8 -> WriteTx.InAnyRecentEra Cardano.Tx -> Property)
```
`prop_bootstrapWitnesses` turned into basically a generator 🤔

Anviking replied:
🤷‍♂️

Anviking commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -460,19 +476,15 @@ import qualified Cardano.Wallet.Primitive.Types.Tx.TxOut.Gen as TxOutGen
 import qualified Data.ByteString.Char8 as B8
 import qualified Data.Foldable as F
+import Data.Generics.Product
+    ( setField )
```
Wrong section

jonathanknowles replied:
A casualty of HLS? 😄

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -758,6 +821,13 @@ fromCardanoTx = \case
+    -> Cardano.Tx era
+toCardanoTx = Cardano.ShelleyTx (shelleyBasedEraFromRecentEra $ recentEra @era)
+
+
```
```suggestion

```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -826,10 +896,22 @@ evaluateMinimumFee era pp tx kwc =
+        intCast $ feePerByte * bytes
+      where
+        bytes :: Natural
+        bytes = sizeOf_BootstrapWitnesses $ intCast nBootstrapWits
```
Suggestion: be a little more explicit that is a _count_ of bytes, rather than the byte values themselves. (I'm being pedantic here, but since we often use `bytes` for byte strings, it's nice to distinguish them at the name level.)

```suggestion
        intCast $ feePerByte * byteCount
      where
        byteCount :: Natural
        byteCount = sizeOf_BootstrapWitnesses $ intCast nBootstrapWits
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -826,10 +896,22 @@ evaluateMinimumFee era pp tx kwc =
+        sizeOf_BootstrapWitnesses 0 = 0
+        sizeOf_BootstrapWitnesses n = 4 + 180 * n
+
+        feePerByte :: Natural
+        feePerByte = case era of
```
Suggestion: you could perform the conversion to `Coin` earlier, which would allow the type of `feePerByte` to reflect that it's an amount of lovelace.

(For the multiplication, you can actually use [`mtimesDefault`](https://hackage.haskell.org/package/base-4.18.0.0/docs/src/Data.Semigroup.html#mtimesDefault) to multiply `Coin` values by `Natural` values, which under the hood is equivalent to `*`).

```suggestion
        feePerByte :: Coin
        feePerByte = Coin.fromNatural $ case era of
```

Anviking replied:
Note that this is the `Ledger` coin type, and there doesn't seem to be a good existing `*`, so I'm sticking to `Natural` for now.

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -422,7 +437,7 @@ import Test.QuickCheck
     ( genericArbitrary, genericShrink )
 import Test.QuickCheck.Extra
-    ( chooseNatural, report )
+    ( chooseNatural, report, shrinkBoundedEnum )
```
This reminds me, I have a PR on `quickcheck` to add this exact function: https://github.com/nick8325/quickcheck/pull/350

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2363,6 +2375,65 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+        let measuredWitSize :: IsCardanoEra era => Cardano.Tx era -> Natural
+            measuredWitSize (Cardano.Tx body wits) = fromIntegral $
+                  serializedSize (Cardano.Tx body wits)
+                - serializedSize (Cardano.Tx body [])
```
To avoid non-standard indents, we can use either:

```suggestion
            measuredWitSize (Cardano.Tx body wits) = fromIntegral
                $ serializedSize (Cardano.Tx body wits)
                - serializedSize (Cardano.Tx body [])
```

or

```suggestion
            measuredWitSize (Cardano.Tx body wits) = fromIntegral $
                serializedSize (Cardano.Tx body wits) -
                serializedSize (Cardano.Tx body [])
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2363,6 +2375,65 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
+
+        let evaluateMinimumFeeDerivedWitSize (Cardano.Tx body wits) =
+                  evaluateMinimumFeeSize (Cardano.Tx body wits)
+                - evaluateMinimumFeeSize (Cardano.Tx body [])
```
```suggestion
        let evaluateMinimumFeeDerivedWitSize (Cardano.Tx body wits)
                = evaluateMinimumFeeSize (Cardano.Tx body wits)
                - evaluateMinimumFeeSize (Cardano.Tx body [])
```

Anviking replied:
TBH, I like the visual prominence of the `-` which currently exists

jonathanknowles replied:
> TBH, I like the visual prominence of the `-` which currently exists

I only mentioned this because _strictly_ speaking it's inconsistent with our coding standards:
https://input-output-hk.github.io/adrestia/code/Coding-Standards#avoid-variable-length-indentation

Other code uses a pattern where everything is aligned at multiple of the indent level (which is consistent with the standard):
```hs
sizeOf_BootstrapWitness
    = sizeOf_SmallArray
    + sizeOf_VKey
    + sizeOf_Signature
    + sizeOf_ChainCode
```

However, I don't feel terribly strongly about it in this particular case.

Anviking replied:
> Other code uses a pattern where everything is aligned at multiple of the indent level (which is consistent with the standard):

Ok, fair enough 👍

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3454,6 +3525,10 @@ shrinkStrictMaybe = \case
 
+instance Arbitrary (Index 'WholeDomain depth) where
+    arbitrary = arbitraryBoundedEnum
+    shrink = shrinkBoundedEnum
```
👍🏻

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,80 @@ prop_balanceTransactionExistingReturnCollateral
+    p n (AnyRecentEra (era :: RecentEra era)) net accIx addr0Ix =
+    let
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indicies.
```
small typo:
```suggestion
        -- around, to ensure we always have 'n' unique indices.
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,80 @@ prop_balanceTransactionExistingReturnCollateral
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indicies.
+        addrIxs = take (fromIntegral n)
+                $ [addr0Ix .. maxBound] ++ [minBound .. addr0Ix]
```
```suggestion
        addrIxs = take (fromIntegral n)
            $ [addr0Ix .. maxBound] ++ [minBound .. addr0Ix]
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,80 @@ prop_balanceTransactionExistingReturnCollateral
+    emptyCardanoTxBody =
+        let
+            Cardano.Tx body _ = WriteTx.toCardanoTx @era $ WriteTx.emptyTx era
+        in body
```
Does the following not work? (It's a line shorter and uses less horizontal space.)
```suggestion
    emptyCardanoTxBody = body
      where
        Cardano.Tx body _ = WriteTx.toCardanoTx @era $ WriteTx.emptyTx era
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,80 @@ prop_balanceTransactionExistingReturnCollateral
+
+(.>=) :: (Show a, Ord a) => a -> a -> Property
+a .>= b = counterexample (show a <> " < " <> show b)
+        $ property $ a >= b
```
Suggestion 1: use `.>=.`, as it follows the existing QC style of defining new versions of existing operators by surrounding them with a `.` on both sides.
Suggestion 2: how about putting this function in `Test.QuickCheck.Extra`?

```suggestion
(.>=.) :: (Show a, Ord a) => a -> a -> Property
a .>=. b = counterexample (show a <> " < " <> show b)
    $ property $ a >= b
```

jonathanknowles reviewed (commented):
Hi @Anviking 

As this is still in draft, just leaving some comments and suggestions.

Anviking committed:
Apply suggestions from code review

Co-authored-by: Jonathan Knowles <jonathan.knowles@iohk.io>

jonathanknowles commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -73,6 +76,7 @@ module Test.QuickCheck.Extra
     , ScaleDiv (..)
     , ScaleMod (..)
 
+
```
```suggestion
```

jonathanknowles commented on lib/test-utils/src/Test/QuickCheck/Extra.hs:
```diff
@@ -642,6 +646,14 @@ verify condition conditionTitle =
+
+(.>=.) :: (Show a, Ord a) => a -> a -> Property
+a .>=. b = counterexample (show a <> " < " <> show b)
+    $ property $ a >= b
```
👍🏻

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -743,6 +758,54 @@ modifyLedgerBody f (Cardano.Tx body keyWits) = Cardano.Tx body' keyWits
                         auxData
                         validity
 
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -743,6 +758,54 @@ modifyLedgerBody f (Cardano.Tx body keyWits) = Cardano.Tx body' keyWits
+    mempty
+    SNothing
+    SNothing
+    SNothing
```
I'm surprised that `cardano-ledger` doesn't define an `emptyTxBody` or equivalent. Perhaps we could make a PR? (Though it'd probably be ages before we could use it. 😄)

Anviking replied:
I think there will be functions we can use once we bump the ledger version, probably like this: https://github.com/input-output-hk/cardano-ledger/blob/9af95e8812224f42f6fe6fdbb8d5b2e8c85e6260/eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs#L487-L502

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -2363,6 +2375,65 @@ balanceTransactionSpec = describe "balanceTransaction" $ do
         $ property prop_balanceTransactionValid
 
+    describe "bootstrap witnesses" $ do
+        -- Used in 'estimateTxSize' in turn used by coin-selection
```
```suggestion
        -- Used in 'estimateTxSize', and in turn used by coin-selection
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,74 @@ prop_balanceTransactionExistingReturnCollateral
+prop_bootstrapWitnesses
+    p n (AnyRecentEra (era :: RecentEra era)) net accIx addr0Ix =
+    let
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indices.
```
Is it always true that `n <= length [minBound .. maxBound]`?

Anviking replied:
We have `n :: Word8` and `Index` with approximately 2^31 or 2^32 elements depending on `derivationType`.

https://github.com/input-output-hk/cardano-wallet/blob/ab5b9b535ea8dcec4c32b6940c5c5b658351ebf2/lib/wallet/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L344-L360

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,74 @@ prop_balanceTransactionExistingReturnCollateral
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indices.
+        addrIxs = take (fromIntegral n)
+            $ [addr0Ix .. maxBound] ++ [minBound .. addr0Ix]
```
Is it possible that `addr0Ix` could appear twice within `addrIxs`?

To be on the safe side, perhaps we could write:

```suggestion
            $ [addr0Ix .. maxBound] ++ filter (< addr0Ix) [minBound .. addr0Ix]
```
(This avoids the use of `pred`, which will fail with a run-time error if `addr0Ix` has no predecessor.)

Anviking replied:
🤔 seems to work 👍(I think I tried something previously which was too slow. Might have been `cycle` though)

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -3992,6 +4067,74 @@ prop_balanceTransactionExistingReturnCollateral
+        Cardano.Tx body _ = WriteTx.toCardanoTx @era $ WriteTx.emptyTx era
+
+    mw = SomeMnemonic $ either (error . show) id
+        (entropyToMnemonic @12 <$> mkEntropy "0000000000000000")
```
I noticed that this definition is already repeated in four different places within `TransactionSpec`.

Perhaps we could factor out the repetition, i.e., by making a top-level dummy mnemonic?

```hs
dummyMnemonic = SomeMnemonic $ either (error . show) id
    (entropyToMnemonic @12 <$> mkEntropy "0000000000000000")
```

Anviking replied:
Done

jonathanknowles reviewed (commented):
Hi @Anviking 

This mostly looks good to me! Though I have a few comments.

Anviking committed:
Avoid theoretical problem with addrIxs definition

Anviking committed:
Factor out helper `dummyMnemonic`

Slight change in test data in the `delegation :: PartialTx` function.
Now the wallet generating the certificates is the same wallet doing the
balancing. This should not affect the test, and the previous choice
doesn't seem to have been intentional.

Anviking committed:
Fixups and address review suggestions

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -997,14 +999,16 @@ evaluateMinimumFee
+      where
+        bytes :: Natural
+        bytes = fromIntegral $ sizeOf_BootstrapWitnesses $ intCast nBootWits
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/src/Cardano/Wallet/Write/Tx.hs:
```diff
@@ -799,6 +810,13 @@ modifyLedgerBody f (Cardano.Tx body keyWits) = Cardano.Tx body' keyWits
+    -> Core.Tx (ShelleyLedgerEra era)
+emptyTx RecentEraConway = Core.mkBasicTx Core.mkBasicTxBody
+emptyTx RecentEraBabbage = Core.mkBasicTx Core.mkBasicTxBody
+emptyTx RecentEraAlonzo = Core.mkBasicTx Core.mkBasicTxBody
```
I did manage to "shorten" this somehow, but not really sure if it's much better:

```suggestion
emptyTx :: RecentEra era -> Core.Tx (ShelleyLedgerEra era)
emptyTx = \case
    RecentEraConway  -> e
    RecentEraBabbage -> e
    RecentEraAlonzo  -> e
  where
    e :: forall e. Core.EraTx e => Core.Tx e
    e = Core.mkBasicTx Core.mkBasicTxBody
```

Anviking replied:
Yes, I think there are qualities to both approaches. Per https://github.com/input-output-hk/cardano-wallet/pull/3773#discussion_r1144713258 I'd like to try out leaning a bit more towards duplication. Changing the internal style should be easy anyway.

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4131,6 +4207,74 @@ prop_balanceTransactionExistingReturnCollateral
+        -- Start incrementing the ixs upward, and if we reach 'maxBound', loop
+        -- around, to ensure we always have 'n' unique indices.
+        addrIxs = take (fromIntegral n)
+            $ [addr0Ix .. maxBound] ++ filter (< addr0Ix) [minBound .. addr0Ix]
```
👍🏻

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4131,6 +4207,74 @@ prop_balanceTransactionExistingReturnCollateral
+                RecentEraBabbage ->
+                    mkByronWitness body net addr (getRawKey addrK, pwd)
+                RecentEraAlonzo  ->
+                    mkByronWitness body net addr (getRawKey addrK, pwd)
```
I keep racking my brains as to how we can reduce this kind of duplication.

I hope we can come up with a decent solution eventually!

Anviking replied:
Yes 👍 

Honestly I think I slightly prefer the duplication to reasoning about specific ledger-internal constraints in multiple places.

I would like to continue the discussion about adding super-class constraints to `IsRecentEra` / `RecentEra` though... I'm not sure the confusing missing-instance error message we mentioned last time actually is a problem, or at least not sure they make things worse than how they now are 🤔

jonathanknowles reviewed (approved):
Hi @Anviking 

This looks fine to me.

Just some small comments.
